feat: 계좌 해지 제한 조건 및 거래 내역 조회 기능 구현

try3982 opened the pull request:
## #️⃣ Issue Number
- #26 
<!--- ex) #이슈번호, #이슈번호 -->

## 📝 요약(Summary)

<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가

## 개발 상세 내역

### ✅ 1. 계좌 해지 조건 처리
- 미결제 대출, 자동이체가 존재하면 계좌 해지 제한
- 잔액이 존재하면 등록된 출금 계좌로 자동 송금
- 출금 계좌 미등록/비활성 시 해지 제한

### ✅ 2. 계좌 해지 후 신규 계좌 개설 제한
- 해지일로부터 1개월 이내인 경우 계좌 개설 불가
- Account의 unregisteredAt 기반으로 검증

### ✅ 3. 거래 내역 조회 기능
- 계좌별 거래내역 조회 시
  - 기간 필터 적용 (기본 1개월)
  - 거래유형 필터 적용 (입금, 출금, 대출, 자동이체 등)
  - 정렬 방식 지원 (최신순, 오래된 순)
- GET /transaction/{accountNumber} 엔드포인트 추가
 
### ✅ 4. ENUM  ERRORCODE추가
- AUTO_TRANSFER_ACTIVE("자동이체가 등록된 계좌입니다."),
- LOAN_EXISTS("미결제 대출이 존재합니다."),
-  AUTO_TRANSFER_EXISTS("자동이체가 등록된 계좌입니다."),
- WITHDRAW_ACCOUNT_NOT_PROVIDED("출금 계좌가 등록되지 않았습니다."),
- WITHDRAW_ACCOUNT_UNMATCH("출금 계좌가 사용자와 일치하지 않습니다."),
- WITHDRAW_ACCOUNT_INACTIVE("출금 계좌가 해지 상태입니다."),
-  CANNOT_CREATE_ACCOUNT_DUE_TO_RECENT_CLOSURE("해지된 계좌가 있어 1개월 내 신규 계좌 개설이 제한됩니다."),

try3982 committed:
feat: 계좌 해지 제한 조건 및 거래 내역 조회 기능 구현

-.미결제 대출 또는 자동이체 등록된 계좌 해지 제한 구현
-.출금 계좌 미등록/오류 시 계좌 해지 실패 처리
-.계좌 해지 후 신규 계좌 개설 제한 기간 설정 기능
-.특정 계좌의 조회 기능 구현

Paransaik commented on src/main/java/com/track/fin/service/AccountService.java:
```diff
@@ -75,19 +77,32 @@ public List<Account> getAccounts(Long userId) {
+        }
+
+        Account withdrawAccount = getAccountByNumber(withdrawAccountNumber);
+
```
두칸씩 개행 주신 이유가 있을까요?

try3982 replied:
수정하겠습니다

Paransaik commented on src/main/java/com/track/fin/service/AccountService.java:
```diff
@@ -102,17 +117,33 @@ private void validateCreateAccount(User user) {
         }
-    }
 
+        boolean hasRecentClosedAccount = accountRepository.findByUser(user).stream()
```
다른 validate와 마찬가지로 외부로 빼는게 좋을 거 같습니다.

또, 좀 더 가독성있게 하는 방법이 없을까요?

Paransaik commented on src/main/java/com/track/fin/service/AutoTransferService.java:
```diff
@@ -0,0 +1,43 @@
+    private final AccountRepository accountRepository;
+    private final TransactionRepository transactionRepository;
+
+    @Transactional(readOnly = true)
```
readOnly를 true 주신 이유가 뭘까요?

try3982 replied:
DB에서 조회만 하고 엔티티의 값을 변경하거나 저장하는 쓰기 작업은 하지 않기 때문에 readOnly = true로 선언하였습니다.

Paransaik commented on src/main/java/com/track/fin/service/TransactionService.java:
```diff
@@ -56,16 +66,50 @@ private Transaction saveAndGetTransaction(TransactionType type, TransactionResul
+        Account account = accountRepository.findByAccountNumber(accountNumber)
+                .orElseThrow(() -> new AccountException(ACCOUNT_NOT_FOUND));
+
+        List<Transaction> transactions = (types == null || types.isEmpty()) ?
```
의도는 알겠으나 가독성이 너무 떨어집니다. 다른 방법은 없을까요?

Paransaik reviewed (commented):
전박적으로 코드 일관성이 굉장히 잘 지켜지고 있습니다.

추가로, 개발 완료 후 결과 확인이 어려워 시간나실 때 동작 결과 한번 시연 부탁드립니다.

Paransaik commented on src/main/java/com/track/fin/service/AccountService.java:
```diff
@@ -75,19 +77,32 @@ public List<Account> getAccounts(Long userId) {
+            transactionService.transfer(userId, closingAccountNumber, withdrawAccountNumber, closingAccount.getBalance());
+        }
+
+        closingAccount.setAccountStatus(CLOSED);
```
날짜 갱신 빠짐

try3982 committed:
refactor: 리뷰반영
-. deleteAccount메서드 개행(코드 컨벤션 수정)
-. 날짜 갱신 추가

try3982 committed:
refactor:hasClosedAccountWithinLastMonth 메서드 추가

try3982 committed:
refactor:

try3982 committed:
feature:프론트에서 사용자가 자기 계좌를 선택해서  활성화된 계좌목록 조회하는 api 구현

try3982 committed:
refactor: 개행

try3982 committed:
refactor: getAccountsByUserId  사용자 계좌 목록 조회 API 추가

try3982 committed:
refactor: conflict 해결

Paransaik commented on src/main/java/com/track/fin/controller/AccountController.java:
```diff
@@ -44,23 +51,65 @@ public DeleteAccount.Response deleteAccount(
     }
 
+
     @GetMapping("/account/{id}")
```
다른 URL과 마찬가지로 URL 수정이 필요해 보입니다.~ 다른 것들도 한번씩 확인 부탁드려요

Paransaik commented on src/main/java/com/track/fin/controller/AccountController.java:
```diff
@@ -44,23 +51,65 @@ public DeleteAccount.Response deleteAccount(
-                                .accountNumber(accountDto.getAccountNumber())
-                                .balance(accountDto.getBalance())
-                                .build())
+        return accountService.getAccounts(userId).stream()
```
객체 생성 방법에서 알려드린 것처럼 static of, from 생성법을 record에도 적용할 수 있는데, 
갹채 생성 책임을 record에 위임 시도해보시겠어요? 

아래 다른 것들도 있습니다.

Paransaik commented on src/main/java/com/track/fin/controller/AccountController.java:
```diff
@@ -44,23 +51,65 @@ public DeleteAccount.Response deleteAccount(
+    @GetMapping("/{accountId}/collateral")
+    public BigDecimal getCollateralRate(
+            @PathVariable Long accountId,
+            @RequestParam("user_id") Long userId
```
여기는 user_id가 카멜케이스로 안바뀌었네요.
URL도 한번 살펴보셔야할 거 같습니다.

Paransaik commented on src/main/java/com/track/fin/controller/TransactionController.java:
```diff
@@ -107,18 +108,18 @@ public WithdrawalRecord withdraw(
 
-    @PostMapping("transaction/cancel")
-    public CancelBalance.Response cancelBalance(
+    @PostMapping("/transaction/cancel")
```
`@PostMapping("/transaction/cancel")` 
-> 
`@PostMapping("/transactions/{transactionId}/cancel")` 

제가 저번에 설명한 RESTful API 설명을 제대로 이해못하신 듯 합니다.
한번 시간내어 설명 드릴까요?

try3982 replied:
수정하겠습니다.다시 복습하겠습니다.

Paransaik commented on src/main/java/com/track/fin/controller/TransactionController.java:
```diff
@@ -128,27 +129,9 @@ public CancelBalance.Response cancelBalance(
     }
 
     @GetMapping("/transaction/{transactionId}")
```
상동

Paransaik commented on src/main/java/com/track/fin/service/AccountService.java:
```diff
@@ -174,4 +176,10 @@ private String generateUniqueAccountNumber() {
         return String.valueOf(1000000000L + Math.random() * 9000000000L);
```
정수를 직접적으로 사용하는 것보단 정적 변수로 빼서 사용하는게 좋아보아보입니다.

Paransaik reviewed (commented):
PR Commit 전 local에서 Commit Squash 해서 origin에 push 부탁드립니다!

try3982 committed:
refactor: 정수-> 상수로 변경하여 사용

try3982 committed:
refactor: 리뷰 반영
-.AccountController record에 객채생성 위임
-. AccountController url 수정
-. FeeController url 수정

try3982 committed:
refactor: 자동이체 등로고 여부 조회 url 작성

try3982 committed:
refactor:

try3982 committed:
refactor:AccountUrl 수정

try3982 committed:
refactor:AccountUrl 수정

try3982 committed:
refactor:AccountUrl 수정

Paransaik committed:
Merge branch 'dev' into feat/close-history-account
